File: lab4/java/PlagiarismDetector.java
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.NoSuchFileException;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Arrays;

/**
 * Command-line tool for detecting similar files.
 */
public class PlagiarismDetector {

    static String[] mapImplementations = new String[] {"list", "bst", "avl"};

    public static void main(String[] args) throws IOException {

        CommandParser parser = new CommandParser("PlagiarismDetector", "Tool for detecting similar files.");
        parser.addArgument("--documents", "-d", "path to directory of documents")
            .makeRequired();
        parser.addArgument("--index", "-i", "use the optimised file index (default: use the slow version)")
            .makeTrueOption();
        parser.addArgument("--map", "-m", "map implementation (default: use key-value list)")
            .setChoices(mapImplementations).setDefault("list");
        parser.addArgument("--ngram", "-n", "ngram size (default: 5)")
            .makeInteger().setDefault(5);
        parser.addArgument("--limit", "-l", "limit the number of similar file pairs (default: 10)")
            .makeInteger().setDefault(10);
        parser.addArgument("--similarity", "-s", "similarity measure (default: 'absolute')")
            .setChoices(SimilarityCalculator.similarityMeasures).setDefault("absolute");

        CommandParser.Namespace options = parser.parseArgs(args);
        String docPath = options.getString("documents");
        boolean useIndex = options.getBoolean("index");
        String useMap = options.getString("map");
        int ngramSize = options.getInteger("ngram");
        int limitResults = options.getInteger("limit");
        String similarityMeasure = options.getString("similarity");

        SimilarityCalculator calculator = new SimilarityCalculator();
        if (useIndex) calculator = new FasterSimilarityCalculator();

        // Initialise the maps.
        if (useMap.equals("avl")) {
            calculator.fileNgrams = new AVLMap<>(() -> new AVLSet<>());
            calculator.ngramIndex = new AVLMap<>(() -> new AVLSet<>());
            calculator.intersections = new AVLMap<>(() -> new AVLSet<>());
        } else if (useMap.equals("bst")) {
            calculator.fileNgrams = new BSTMap<>(() -> new BSTSet<>());
            calculator.ngramIndex = new BSTMap<>(() -> new BSTSet<>());
            calculator.intersections = new BSTMap<>(() -> new BSTSet<>());
        } else {
            calculator.fileNgrams = new ListMap<>(() -> new ListSet<>());
            calculator.ngramIndex = new ListMap<>(() -> new ListSet<>());
            calculator.intersections = new ListMap<>(() -> new ListSet<>());
        }
    
        // Find all .txt files in the directory and sort the filenames.
        // Uses the Java 8 streams API - very handy Java feature which we don't cover in the course.
        // Read more about streams in 'SimilarityCalculator.findMostSimilar'
        Path[] paths = Files.list(Paths.get(docPath))
            .filter((Path p) -> p.toString().endsWith(".txt"))
            .sorted()
            .toArray(Path[]::new);
        if (paths.length == 0) 
            throw new NoSuchFileException("Empty directory");

        // Create stopwatches to time the execution of each phase of the program.
        Stopwatch stopwatchTotal = new Stopwatch();
        Stopwatch stopwatch = new Stopwatch();

        // Phase 1: Read n-grams from all input files.
        calculator.readNgramsFromFiles(paths, ngramSize);
        calculator.fileNgrams.validate();
        stopwatch.finished("Reading all input files");

        // Phase 2: Build index of n-grams.
        calculator.buildNgramIndex();
        calculator.ngramIndex.validate();
        stopwatch.finished("Building n-gram index");

        // Phase 3: Compute the n-gram intersections of all file pairs.
        calculator.computeIntersections();
        calculator.intersections.validate();
        stopwatch.finished("Computing intersections");

        // Phase 4: Find the L most similar file pairs, arranged in decreasing order of similarity.
        PathPair[] mostSimilar = calculator.findMostSimilar(limitResults, similarityMeasure);
        stopwatch.finished("Finding the most similar files");

        stopwatchTotal.finished("In total the program");
        
        // Print out some statistics.
        System.out.println();
        System.out.println("Balance statistics:");
        System.out.println("  fileNgrams: " + calculator.fileNgrams);
        System.out.println("  ngramIndex: " + calculator.ngramIndex);
        System.out.println("  intersections: " + calculator.intersections);
            
        // Print out the plagiarism report!
        System.out.println();
        System.out.println("Plagiarism report:");
        for (String measure : SimilarityCalculator.similarityMeasures)
            System.out.format("%10s", measure.contains("-") ? measure.replaceAll("-.*", "") : "");
        System.out.println();
        for (String measure : SimilarityCalculator.similarityMeasures)
            System.out.format("%10s", measure.replaceAll(".*-", ""));
        System.out.println();
        int maxFilenameSize = Arrays.stream(mostSimilar)
            .mapToInt((PathPair p) -> p.path1.getFileName().toString().length())
            .max().getAsInt();
        for (PathPair pair : mostSimilar) {
            for (String measure : SimilarityCalculator.similarityMeasures) {
                int decimals = measure.equals("absolute") ? 0 : measure.contains("weighted") ? 2 : 3;
                System.out.format("%10." + decimals + "f", calculator.similarity(pair, measure));
            }
            System.out.format("  %-" + maxFilenameSize + "s %s%n", pair.path1.getFileName(), pair.path2.getFileName());
      }
    }

}

File: lab4/python/plagiarism_detector.py
"""
Command-line tool for detecting similar files.
"""

from pathlib import Path
from typing import Dict, Any, Tuple, Type

from simple_map import SimpleMap
from list_map import ListMap
from bst_map import BSTMap
from avl_map import AVLMap
from simple_set import SimpleSet
from list_set import ListSet
from bst_set import BSTSet
from avl_set import AVLSet
from simliarity_calculator import SimilarityCalculator
from faster_simliarity_calculator import FasterSimilarityCalculator
from command_parser import CommandParser
from stopwatch import Stopwatch


mapImplementations: Dict[str, Tuple[Type[SimpleMap[Any, Any]], Type[SimpleSet[Any]]]] = {
    "list": (ListMap, ListSet),
    "bst": (BSTMap, BSTSet),
    "avl": (AVLMap, AVLSet),
}

parser = CommandParser(description=__doc__.strip())
parser.add_argument("--documents", "-d", required=True, type=Path,
                    help="path to directory of documents")
parser.add_argument("--index", "-i", action="store_true",
                    help="use the optimised file index (default: use the slow version)")
parser.add_argument("--map", "-m", choices=list(mapImplementations), default="list",
                    help="map implementation (default: use key-value list)")
parser.add_argument("--ngram", "-n", type=int, default=5,
                    help="ngram size (default: 5)")
parser.add_argument("--limit", "-l", type=int, default=10,
                    help="limit the number of similar file pairs (default: 10)")
parser.add_argument("--similarity", "-s", choices=SimilarityCalculator.similarityMeasures, default='absolute',
                    help="similarity measure (default: 'absolute')")


def main():
    options = parser.parse_args()
    if not options.documents.is_dir():
        raise FileNotFoundError(f"The directory '{options.documents}' does not exist")

    calculator = SimilarityCalculator()
    if options.index: calculator = FasterSimilarityCalculator()

    # Initialise the maps.
    mapClass, setClass = mapImplementations[options.map]
    calculator.fileNgrams = mapClass(setClass)
    calculator.ngramIndex = mapClass(setClass)
    calculator.intersections = mapClass(setClass)

    # Find all .txt files in the directory and sort the filenames.
    paths = sorted(options.documents.glob('*.txt'))
    if not paths:
        raise FileNotFoundError("Empty directory")

    # Create stopwatches time the execution of each phase of the program.
    stopwatchTotal = Stopwatch()
    stopwatch = Stopwatch()

    # Phase 1: Read n-grams from all input files.
    calculator.readNgramsFromFiles(paths, options.ngram)
    calculator.fileNgrams.validate()
    stopwatch.finished("Reading all input files")

    # Phase 2: Build index of n-grams.
    calculator.buildNgramIndex()
    calculator.ngramIndex.validate()
    stopwatch.finished("Building n-gram index")

    # Phase 3: Compute the n-gram intersections of all file pairs.
    calculator.computeIntersections()
    calculator.intersections.validate()
    stopwatch.finished("Computing intersections")

    # Phase 4: Find the L most similar file pairs, arranged in decreasing order of similarity.
    mostSimilar = calculator.findMostSimilar(options.limit, measure=options.similarity)
    stopwatch.finished("Finding the most similar files")

    stopwatchTotal.finished("In total the program")

    # Print out some statistics.
    print()
    print("Balance statistics:")
    print(f"  fileNgrams: {calculator.fileNgrams}")
    print(f"  ngramIndex: {calculator.ngramIndex}")
    print(f"  intersections: {calculator.intersections}")

    # Print out the plagiarism report!
    print()
    print("Plagiarism report:")
    print("".join(f"{measure.rpartition('-')[0]:>10s}" for measure in calculator.similarityMeasures))
    print("".join(f"{measure.rpartition('-')[-1]:>10s}" for measure in calculator.similarityMeasures))
    maxFilenameSize = max(len(pair[0].name) for pair in mostSimilar)
    for pair in mostSimilar:
        for measure in calculator.similarityMeasures:
            decimals = 0 if measure == 'absolute' else 2 if 'weighted' in measure else 3
            print(f"{calculator.similarity(pair, measure):>10.{decimals}f}", end="")
        print(f"  {pair[0].name:{maxFilenameSize}s} {pair[1].name:s}")


if __name__ == '__main__':
    main()


